tests/utils/test_rotation_utils.py

```diff
@@ -0,0 +1,148 @@
+    """Return data for a scenario with 5 camera poses, with ordering that does NOT follow their connectivity.
+
+    Below, we do NOT specify i1 < i2 for all edges (i1,i2).
+
+    Graph topology:
+
+              | 3     | 0
+              o-- ... o--
+              .       .
+              .       .
+    |         |       |
+    o-- ... --o     --o
+    4         1       2
+
+    """
+    # Expected angles.
+    wRi_list_euler_deg_expected = [0,90,90,0,0]
+
+    # Ground truth 2d rotations for 5 ordered poses (0,1,2,3,4)
+    wRi_list_gt = [Rot3.RzRyRx(np.deg2rad(Rz_deg), 0, 0) for Rz_deg in wRi_list_euler_deg_expected]
+
+    edges = [(1, 4), (1, 3), (0, 3), (0, 2)]
+    i2Ri1_dict = _create_synthetic_relative_pose_measurements(wRi_list_gt=wRi_list_gt, edges=edges)
+
+    return i2Ri1_dict, wRi_list_euler_deg_expected
+
+
+def _create_synthetic_relative_pose_measurements(
+    wRi_list_gt: List[Rot3], edges: List[Tuple[int, int]]
+) -> Dict[Tuple[int, int], Rot3]:
+    """Generate synthetic relative rotation measurements, from ground truth global rotations.
+
+    Args:
+        wRi_list_gt: List of (3,3) rotation matrices.
+        edges: Edges as pairs of image indices.
+
+    Returns:
+        i2Ri1_dict: Relative rotation measurements.
+    """
+    i2Ri1_dict = {}
+    for i1, i2 in edges:
+        wRi2 = wRi_list_gt[i2]
+        wRi1 = wRi_list_gt[i1]
+        i2Ri1_dict[(i1, i2)] = wRi2.inverse() * wRi1
+
+    return i2Ri1_dict
+
+
+class TestRotationUtil(unittest.TestCase):
+    def test_mst_initialization(self):
+        """Test for 4 poses in a circle, with a pose connected all others."""
+        i2Ri1_dict, wRi_expected = sample_poses.convert_data_for_rotation_averaging(
+            sample_poses.CIRCLE_ALL_EDGES_GLOBAL_POSES, sample_poses.CIRCLE_ALL_EDGES_RELATIVE_POSES
+        )
+
+        wRi_computed = rotation_util.initialize_global_rotations_using_mst(
+            len(wRi_expected),
+            i2Ri1_dict,
+            edge_weights={(i1, i2): (i1 + i2) * 100 for i1, i2 in i2Ri1_dict.keys()},
+        )
+        self.assertTrue(
+            geometry_comparisons.compare_rotations(wRi_computed, wRi_expected, ROTATION_ANGLE_ERROR_THRESHOLD_DEG)
+        )
+
+    def test_greedily_construct_st_ordered_chain(self) -> None:
+        """Ensures that we can greedily construct a Spanning Tree for an ordered chain."""
+
+        i2Ri1_dict, wRi_list_euler_deg_expected = _get_ordered_chain_pose_data()
+
+        num_images = 5
+        wRi_list_computed = rotation_util.initialize_global_rotations_using_mst(
+            num_images,
+            i2Ri1_dict,
+            edge_weights={(i1, i2): (i1 + i2) * 100 for i1, i2 in i2Ri1_dict.keys()},
+        )
+
+        wRi_list_euler_deg_est = [np.rad2deg(wRi.roll()) for wRi in wRi_list_computed]
+        assert np.allclose(wRi_list_euler_deg_est, wRi_list_euler_deg_expected)
+
+    def test_greedily_construct_st_mixed_order_chain(self) -> None:
```

Comment:
@ayushbaid I added a test that fails with the current implementation

Reply:
Do we need to change the lookup logic? Or do you have a fix in mind?

Reply:
I added a fix -- traversing the shortest path in the MST between the origin and destination, and chaining the poses together